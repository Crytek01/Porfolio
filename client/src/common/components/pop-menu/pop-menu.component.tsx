import { FunctionComponent } from 'react';
import { ILinkOptions } from '@common/components/link/link.component';

import {
  MenuItem,
  PopLink,
  PopLinkName,
  PopMenuContainer,
} from './pop-menu.styles';

export interface IMenuOptions {
  name: string;
  route: string;
  isExternal?: boolean;
  target?: ILinkOptions['target'];
}

interface IPopMenu {
  isMenuOpen: boolean;
  onItemClick: () => void;
  menuColor?: string;
  menuFontColor?: string;
  options: readonly IMenuOptions[];
}

export const PopMenu: FunctionComponent<IPopMenu> = ({
  isMenuOpen,
  onItemClick,
  menuColor,
  options,
}) => (
  <PopMenuContainer
    backgroundColor={menuColor}
    onClick={(e) => e.stopPropagation()}
    isVisible={isMenuOpen}
  >
    {options.map(({ route, name, isExternal, target = '_self' }) => (
      <MenuItem key={route}>
        <PopLink
          to={route}
          onClick={onItemClick}
          isExternal={isExternal}
          options={{ target }}
        >
          <PopLinkName>{name}</PopLinkName>
        </PopLink>
      </MenuItem>
    ))}
  </PopMenuContainer>
);
