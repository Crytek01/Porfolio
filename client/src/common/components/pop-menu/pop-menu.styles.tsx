import { CustomLink } from '@common/components/custom-link/custom-link.component';
import styled, { css } from 'styled-components';

import { styleSize, Typography } from '../typography/typography.component';

const hideElementCSS = css`
  opacity: 0;
  pointer-events: none;
  visibility: hidden;
`;

interface IPopMenuContainer {
  backgroundColor?: string;
  isVisible: boolean;
}

export const PopMenuContainer = styled.ul<IPopMenuContainer>`
  transition: all 0.3s ease-in;

  background-color: ${({ backgroundColor = '#fff' }) => backgroundColor};
  border: 1px solid #333;
  list-style: none;

  box-shadow: inset 0 0.8rem 4.8rem rgb(0 0 0 / 3%);

  overflow: hidden;

  border-radius: 9px;

  display: flex;
  flex-direction: column;

  align-items: center;
  justify-content: start;

  position: absolute;
  top: 55px;
  right: 10px;

  width: 20rem;

  padding: 0.2rem 0;

  ${({ isVisible }) => (isVisible ? null : hideElementCSS)}
`;

export const MenuItem = styled.li`
  display: block;
  font-size: ${({ theme: { fontSize } }) => fontSize[3]};
  padding: 0.5rem 1rem;
  width: 100%;

  &:not(:last-child) {
    border-bottom: 1px solid #d3d3d3;
  }

  &:focus {
    box-shadow: 0 0 0 0.4rem rgba(0, 204, 255, 0.5);
  }
`;

export const PopLink = styled(CustomLink)`
  display: flex;
  align-items: center;
  justify-content: start;
  gap: ${({ theme: { spacing } }) => spacing[1]};

  padding: 1rem;
`;

export const PopLinkName = styled(Typography).attrs({
  size: 'b3',
  color: 'semiBoldGrey',
  weight: 'medium',
  align: 'left',
})`
  @media (max-width: 19em) {
    font-size: ${styleSize.b4};
  }
`;
