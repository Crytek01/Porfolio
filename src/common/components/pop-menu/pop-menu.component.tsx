import { FunctionComponent } from 'react';

import {
  MenuItem,
  PopLink,
  PopLinkName,
  PopMenuContainer,
} from './pop-menu.styles';

export interface IMenuOptions {
  name: string;
  route: string;
}

interface IPopMenu {
  isMenuOpen: boolean;
  onItemClick: () => void;
  menuColor?: string;
  menuFontColor?: string;
  options: readonly IMenuOptions[];
}

export const PopMenu: FunctionComponent<IPopMenu> = ({
  isMenuOpen,
  onItemClick,
  menuColor,
  options,
}) => (
  <PopMenuContainer
    backgroundColor={menuColor}
    onClick={(e) => e.stopPropagation()}
    isVisible={isMenuOpen}
  >
    {options.map(({ route, name }) => (
      <MenuItem key={route}>
        <PopLink to={route} onClick={onItemClick}>
          <PopLinkName>{name}</PopLinkName>
        </PopLink>
      </MenuItem>
    ))}
  </PopMenuContainer>
);
