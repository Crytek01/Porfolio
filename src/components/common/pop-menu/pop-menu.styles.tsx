import styled, { css } from 'styled-components';

const hideElementCSS = css`
  opacity: 0;
  pointer-events: none;
  visibility: hidden;
`;

interface IPopMenuContainer {
  backgroundColor?: string;
  isVisible: boolean;
}

export const PopMenuContainer = styled.ul<IPopMenuContainer>`
  transition: all 0.8s ease-in;

  background-color: ${({ backgroundColor = '#888' }) => backgroundColor};
  list-style: none;

  box-shadow: inset 0 0.8rem 4.8rem rgb(0 0 0 / 3%);

  overflow: hidden;

  border-radius: 9px;

  display: flex;
  flex-direction: column;

  align-items: flex-start;
  justify-content: start;

  position: absolute;
  top: 50px;
  right: 10px;

  z-index: 999;

  width: 20rem;

  padding: 1.2rem 0;

  ${({ isVisible }) => (isVisible ? null : hideElementCSS)}
`;

export const MenuItem = styled.li`
  display: block;
  font-size: 1.6rem;
  padding: 0 2rem;

  &:focus {
    box-shadow: 0 0 0 0.4rem rgba(0, 204, 255, 0.5);
  }
`;
