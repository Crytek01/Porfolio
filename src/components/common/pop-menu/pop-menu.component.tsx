import { FunctionComponent } from 'react';
import { RouterLink } from '../nav-bar/nav-bar.styled';
import { MenuItem, PopMenuContainer } from './pop-menu.styles';

export interface IMenuOptions {
  name: string;
  route: string;
}

interface IPopMenu {
  isMenuOpen: boolean;
  onItemClick: () => void;
  menuColor?: string;
  menuFontColor?: string;
  options: readonly IMenuOptions[];
}

export const PopMenu: FunctionComponent<IPopMenu> = ({
  isMenuOpen,
  onItemClick,
  menuColor,
  menuFontColor,
  options,
}) => (
  <PopMenuContainer
    backgroundColor={menuColor}
    onClick={(e) => e.stopPropagation()}
    isVisible={isMenuOpen}
  >
    {options.map(({ route, name }) => (
      <MenuItem key={route}>
        <RouterLink to={route} onClick={onItemClick}>
          {name}
        </RouterLink>
      </MenuItem>
    ))}
  </PopMenuContainer>
);
