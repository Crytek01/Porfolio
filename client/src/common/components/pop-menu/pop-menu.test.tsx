import { shallowRender } from '@common/utils/test/shallowRender';
import { describe, expect, it } from 'vitest';

import { PopMenu } from './pop-menu.component';

describe('Pop menu component Test', () => {
  it('renders correctly', () => {
    const renderer = shallowRender(
      <PopMenu
        isMenuOpen={false}
        onItemClick={function (): void {
          console.log();
        }}
        options={[
          {
            name: 'Test1',
            route: '/test1',
          },
          {
            name: 'Test2',
            route: '/test2',
          },
          {
            name: 'Test3',
            route: '/test3',
            isExternal: true,
            target: '_blank',
          },
        ]}
      />
    );

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
