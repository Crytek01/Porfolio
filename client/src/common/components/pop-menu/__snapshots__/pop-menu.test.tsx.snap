// Vitest Snapshot v1

exports[`Pop menu component Test > renders correctly 1`] = `
<styled.ul
  isVisible={false}
  onClick={[Function]}
>
  <styled.li>
    <Styled(Link)
      onClick={[Function]}
      options={
        {
          "target": "_self",
        }
      }
      to="/test1"
    >
      <Styled(Typography)>
        Test1
      </Styled(Typography)>
    </Styled(Link)>
  </styled.li>
  <styled.li>
    <Styled(Link)
      onClick={[Function]}
      options={
        {
          "target": "_self",
        }
      }
      to="/test2"
    >
      <Styled(Typography)>
        Test2
      </Styled(Typography)>
    </Styled(Link)>
  </styled.li>
  <styled.li>
    <Styled(Link)
      isExternal={true}
      onClick={[Function]}
      options={
        {
          "target": "_blank",
        }
      }
      to="/test3"
    >
      <Styled(Typography)>
        Test3
      </Styled(Typography)>
    </Styled(Link)>
  </styled.li>
</styled.ul>
`;
